src/backend/distributed/commands/utility_hook.c

```diff
@@ -647,6 +691,134 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 }
 
 
+/*
+ * UndistributeDisconnectedCitusLocalTables undistributes citus local tables that
+ * are not connected to any reference tables via their individual foreign key
+ * subgraphs.
+ */
+static void
+UndistributeDisconnectedCitusLocalTables(void)
+{
+	List *citusLocalTableIdList = CitusTableTypeIdList(CITUS_LOCAL_TABLE);
+	citusLocalTableIdList = SortList(citusLocalTableIdList, CompareOids);
+
+	/* acquire ExclusiveLock as UndistributeTable does */
+	LockRelationsWithLockMode(citusLocalTableIdList, ExclusiveLock);
```

Comment:
why do we need such an aggressive lock if we're just doing a few checks?

Maybe RowExclusiveLock to conflict with foreign key creation. Also seems like we could release locks if nothing happens / just keep whatever locks UndistributeTable takes.

Reply:
both makes sense, (iirc, `ShareRowExclusiveLock` is the lock that we should acquire for concurrent fkey creation)